File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.cpp
class Solution {
public:
    string reverseVowels(string s) {
        int n = s.length();
        auto isVowel = [](char ch) {
            ch = tolower(ch);
            return (ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u');
        };
        int start = 0, end = n-1;
        while(start<end){
            if(isVowel(s[start]) && isVowel(s[end])){
                swap(s[start],s[end]);
                start++;
                end--;
            }
            if(!isVowel(s[start])){
                start++;
            }
            if(!isVowel(s[end])){
                end--;
            }
        }
        return s;
    }
};

// A E A E I O Z Z I O T T

// A E A E I O I O
// O I O I E A E A

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.java
class Solution {
    private void swap(char[] word,int start,int end){
        char temp = word[start];
        word[start] = word[end];
        word[end] = temp;
    }
    public String reverseVowels(String s) {
        char[] word = s.toCharArray();
        int n = s.length();
        Predicate<Character> isVowel = c -> "aeiouAEIOU".indexOf(c) != -1;
        int start = 0, end = n-1;
        while(start<end){
            boolean isStartVowel = isVowel.test(word[start]);
            boolean isEndVowel = isVowel.test(word[end]);
            
            if(isStartVowel && isEndVowel){
                swap(word,start,end);
                start++;
                end--;
            }
            if(!isStartVowel){
                start++;
            }
            if(!isEndVowel){
                end--;
            }
        }
        return new String(word);
    }
}